eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/JpaTransactionalExecutorProvider.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2010-2025. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.eventsourcing.eventstore.jpa;
+
+import jakarta.annotation.Nonnull;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.EntityTransaction;
+import org.axonframework.common.annotation.Internal;
+import org.axonframework.common.function.ThrowingFunction;
+import org.axonframework.common.jpa.EntityManagerExecutor;
+import org.axonframework.common.tx.TransactionalExecutor;
+import org.axonframework.eventsourcing.eventstore.TransactionalExecutorProvider;
+import org.axonframework.messaging.core.Context.ResourceKey;
+import org.axonframework.messaging.core.unitofwork.ProcessingContext;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+/**
+ * A {@link TransactionalExecutorProvider} implementation for JPA {@link EntityManager}s which
+ * provides a {@link TransactionalExecutor}.
+ * <p>
+ * When a processing context is supplied, supplies the {@link TransactionalExecutor} it must contain.
+ * If no processing context is supplied, creates an executor that executes the supplied functions
+ * in their own transaction.
+ *
+ * @author John Hendrikx
+ * @since 5.1.0
+ */
+@Internal
+public class JpaTransactionalExecutorProvider implements TransactionalExecutorProvider<EntityManager> {
+
+    /**
+     * The resource key for the {@link EntityManagerExecutor} supplier.
+     */
+    public static final ResourceKey<Supplier<EntityManagerExecutor>> SUPPLIER_KEY = ResourceKey.withLabel(EntityManagerExecutor.class.getSimpleName());
+
+    private final EntityManagerFactory entityManagerFactory;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param entityManagerFactory an {@link EntityManager} used when no processing context is supplied.
+     */
+    public JpaTransactionalExecutorProvider(@Nonnull EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = Objects.requireNonNull(entityManagerFactory, "entityManagerFactory");
+    }
+
+    @Override
+    public TransactionalExecutor<EntityManager> getTransactionalExecutor(ProcessingContext processingContext) {
```

Comment:
Not sure if Eclipse does this, but Intellij provides warnings if overridden methods their parameters are not annotated as they are on the interface:

```suggestion
    public TransactionalExecutor<EntityManager> getTransactionalExecutor(@Nullable ProcessingContext processingContext) {
```